sdv/multi_table/utils.py

```diff
@@ -0,0 +1,431 @@
+"""Utility functions for the MultiTable models."""
+import math
+from collections import defaultdict
+from copy import deepcopy
+
+import numpy as np
+import pandas as pd
+
+from sdv._utils import _get_root_tables
+from sdv.multi_table import HMASynthesizer
+from sdv.multi_table.hma import MAX_NUMBER_OF_COLUMNS
+
+MODELABLE_SDTYPE = ['categorical', 'numerical', 'datetime', 'boolean']
+
+
+def _get_relationship_for_child(relationships, child_table):
+    return [rel for rel in relationships if rel['child_table_name'] == child_table]
+
+
+def _get_relationship_for_parent(relationships, parent_table):
+    return [rel for rel in relationships if rel['parent_table_name'] == parent_table]
+
+
+def _get_n_order_descendants(relationships, parent_table, order):
+    """Get the descendants of the parent table until a given order.
+
+    Args:
+        relationships (list[dict]):
+            List of relationships between the tables.
+        parent_table (str):
+            Name of the parent table.
+        order (int):
+            Order of the descendants.
+    """
+    descendants = {}
+    order_1_descendants = _get_relationship_for_parent(relationships, parent_table)
+    descendants['order_1'] = [rel['child_table_name'] for rel in order_1_descendants]
+    for i in range(2, order+1):
+        descendants[f'order_{i}'] = []
+        prov_descendants = []
+        for child_table in descendants[f'order_{i-1}']:
+            order_i_descendants = _get_relationship_for_parent(relationships, child_table)
+            prov_descendants.extend([rel['child_table_name'] for rel in order_i_descendants])
+
+        descendants[f'order_{i}'] = prov_descendants
+
+    return descendants
+
+
+def _get_all_descendant_per_root_at_order_n(relationships, order):
+    """Get all the descendants of the root tables at a given order.
+
+    Args:
+        relationships (list[dict]):
+            List of relationships between the tables.
+        order (int):
+            Order of the descendants.
+    """
+    root_tables = _get_root_tables(relationships)
+    all_descendants = {}
+    for root in root_tables:
+        order_descendants = _get_n_order_descendants(relationships, root, order)
+        all_descendant_root = set()
+        for order_key in order_descendants:
+            all_descendant_root.update(order_descendants[order_key])
+
+        all_descendants[root] = all_descendant_root
+
```

Comment:
could we prevent having to recompute children and grandchildren later by having this method return a dict in the following format:
```
root -> {'order_1': ..., 'order_2': ..., 'num_descendants': 6}
```

Reply:
Good idea, done in [209b720](https://github.com/sdv-dev/SDV/pull/1874/commits/209b720e703f9fef6eb4ceb112c5506804198b66)